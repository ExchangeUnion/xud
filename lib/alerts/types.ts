import { AlertType, ChannelSide } from '../constants/enums';

type BaseAlert = {
  type: AlertType;
  message: string;
  date: number;
};

export type BalanceAlertEvent = {
  /** The total balance of the channel when the alert is triggered. */
  totalBalance: number;
  /** The side of the balance either local or remote. */
  side: ChannelSide;
  /** The balance that triggered the alert. */
  sideBalance: number;
  /** The alert threshold in percentage, e.g. 10 means %10. */
  bound: number;
  /** The currency of the channel. */
  currency: string;
};
export type BalanceAlert = BaseAlert & BalanceAlertEvent;

export type Alert = BalanceAlert;
