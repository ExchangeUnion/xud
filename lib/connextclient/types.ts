/**
 * The configurable options for the connext client.
 */
export type ConnextClientConfig = {
  disable: boolean;
  host: string;
  port: number;
  webhookhost: string;
  webhookport: number;
};

/** General information about the state of this connext client. */
export type ConnextInfo = {
  status: string;
  error?: string;
  address?: string;
  chain?: string;
  version?: string;
};

/** Response for eth_blockNumber call  */
export type ConnextBlockNumberResponse = {
  result: string;
};

/** Response for eth_gasPrice call  */
export type EthproviderGasPriceResponse = {
  result: string;
};

type Balance = {
  amount: string[];
  to: string[];
};

/** Connext channel details */
export type ConnextChannelDetails = {
  assetIds: string[];
  balances: Balance[];
  channelAddress: string;
  merkleRoot: string;
  processedDepositsA: string[];
  processedDepositsB: string[];
  networkContext: {
    chainId: number;
    channelFactoryAddress: string;
    transferRegistryAddress: string;
    providerUrl: string;
  };
  nonce: number;
  alice: string;
  aliceIdentifier: string;
  bob: string;
  bobIdentifier: string;
  timeout: string;
  latestUpdate: {
    assetId: string;
    balance: Balance;
    channelAddress: string;
    details: {
      totalDepositsAlice: string;
      totalDepositsBob: string;
    };
    fromIdentifier: string;
    nonce: number;
    aliceSignature: string;
    bobSignature: string;
    toIdentifier: string;
    type: string;
  };
  defundNonce: string;
  inDispute: boolean;
};

/**
 * The payload for tokenPayment call.
 */
export type ConnextTransferRequest = {
  type: 'HashlockTransfer';
  channelAddress: string;
  amount: string;
  assetId: string;
  details: {
    lockHash: string;
    expiry: string;
  };
  recipient: string;
  meta: {
    routingId: string;
  };
  publicIdentifier: string;
};

/**
 * The response for errors.
 */
export type ConnextErrorResponse = { message: string };

export type ConnextConfig = {
  publicIdentifier: string;
  signerAddress: string;
  index: number;
};

/**
 * The response for /config call.
 */
export type ConnextConfigResponse = ConnextConfig[];

/**
 * The response for /channel call.
 */
export type ConnextChannelResponse = string[];

/**
 * The response for balance call.
 */
export type ConnextChannelBalanceResponse = {
  freeBalanceOffChain: string;
  nodeFreeBalanceOffChain: string;
  freeBalanceOnChain: string;
};

export type GetBlockByNumberResponse = {
  result: {
    difficulty: string;
    extraData: string;
    gasLimit: string;
    gasUsed: string;
    hash: string;
    logsBloom: string;
    miner: string;
    mixHash: string;
    nonce: string;
    number: string;
    parentHash: string;
    receiptsRoot: string;
    sha3Uncles: string;
    size: string;
    stateRoot: string;
    timestamp: string;
    totalDifficulty: string;
    transactions: string[];
    transactionsRoot: string;
    uncles: string[];
  };
};

/**
 * The response for hashLockTransfer call.
 */
export type ConnextTransferResponse = {
  channelAddress: string;
  transferId: string;
  routingId: string;
};

/**
 * The response for withdraw call.
 */
export type ConnextWithdrawResponse = {
  channelAddress: '0xa929dB0530daB525596f5d48Fb5C322fDa8A337B';
  transferId: '0xc2e4592d3fb6c02ee1d3b07bed83b5914f8ca084b0f91d6b80bf8107e58c9c38';
};

type ConnextRoutingPath = {
  recipient: string;
  recipientChainId: number;
  recipientAssetId: string;
};

export type ConnextTransfer = {
  channelFactoryAddress: string;
  assetId: string;
  chainId: number;
  channelAddress: string;
  balance: {
    amount: string[];
    to: string[];
  };
  initiator: string;
  responder: string;
  initialStateHash: string;
  transferDefinition: string;
  transferEncodings: string[];
  transferId: string;
  transferState: {
    lockHash: string;
    expiry: string;
  };
  transferTimeout: string;
  meta: {
    requireOnline: boolean;
    routingId: string;
    path: ConnextRoutingPath[];
    senderIdentifier?: string;
  };
  transferResolver?: {
    preImage?: string;
  };
};

export type TransfersByRoutingIdResponse = ConnextTransfer[];

export type ExpectedIncomingTransfer = {
  rHash: string;
  units: bigint;
  expiry: number;
  tokenAddress: string;
  routingId: string;
};

export type ConnextPreimageRequest = {
  channelAddress: string;
  channelBalance: {
    to: string[];
    amount: string[];
  };
  transfer: {
    channelFactoryAddress: string;
    assetId: string;
    chainId: string;
    channelAddress: string;
    balance: {
      amount: string[];
      to: string[];
    };
    initiator: string;
    responder: string;
    initialStateHash: string;
    transferDefinition: string;
    transferEncodings: string[];
    transferId: string;
    transferState: {
      lockHash: string;
      expiry: string;
    };
    transferTimeout: string;
    meta: {
      requireOnline: boolean;
      routingId: string;
      path: [
        {
          recipient: string;
          recipientChainId: number;
          recipientAssetId: string;
        },
      ];
    };
    transferResolver: {
      preImage: string;
    };
  };
  conditionType: string;
};

export type ConnextIncomingTransferRequest = {
  channelAddress: string;
  channelBalance: {
    to: string[];
    amount: string[];
  };
  transfer: {
    channelFactoryAddress: string;
    assetId: string;
    chainId: number;
    channelAddress: string;
    balance: {
      amount: string[];
      to: string[];
    };
    initiator: string;
    responder: string;
    initialStateHash: string;
    transferDefinition: string;
    transferEncodings: string[];
    transferId: string;
    transferState: {
      lockHash: string;
      expiry: string;
    };
    transferTimeout: string;
    meta: {
      routingId: string;
    };
  };
  conditionType: string;
};

export type ConnextDepositConfirmedRequest = {
  id: string;
  data?: {
    amount: {
      _hex: string;
    };
    assetId: string;
    hash: string;
  };
};

export type ProvidePreimageEvent = {
  rHash: string;
  preimage: string;
};

export type TransferReceivedEvent = {
  tokenAddress: string;
  rHash: string;
  expiry: number;
  units: bigint;
  routingId: string;
};
