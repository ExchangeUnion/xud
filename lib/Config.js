const fs = require('fs');
const os = require('os');
const toml = require('toml');
const util = require('./utils/utils');
const enums = require('./constants/enums');


class Config {
  constructor(args) {
    this.args = args || null;
    const platform = os.platform();
    let lndDatadir;
    switch (platform) {
      case 'win32': { // windows
        const homeDir = process.env.LOCALAPPDATA;
        this.xuDir = `${homeDir}/Xud/`;
        lndDatadir = `${homeDir}/Lnd/`; // default lnd directory location
        break;
      }
      case 'darwin': { // mac
        const homeDir = process.env.HOME;
        this.xuDir = `${homeDir}/.xud/`;
        lndDatadir = `${homeDir}/Library/Application Support/Lnd/`;
        break;
      }
      default: { // linux
        const homeDir = process.env.HOME;
        this.xuDir = `${homeDir}/.xud/`;
        lndDatadir = `${homeDir}/.lnd/`;
        break;
      }
    }

    // default configuration
    this.p2p = {
      listen: true,
      port: 8885, // X = 88, U = 85 in ASCII
    };
    this.db = {
      host: 'localhost',
      port: 3306,
      username: 'xud',
      password: null,
      database: 'xud',
      dialect: 'mysql',
      pool: { max: 100 },
      operatorsAliases: false,
    };
    this.testDb = {
      ...this.db,
      database: 'xud_test',
    };
    this.raiden = {
      port: 5001,
    };
    this.rpcPort = 8886;
    this.lnd = {
      datadir: lndDatadir,
      rpcProtoPath: 'lndrpc.proto',
    };
    this.swapProtocols = {
      [enums.swapProtocols.LND]: {
        enable: true,
      },
      [enums.swapProtocols.RAIDEN]: {
        enable: true,
      },
    };
  }

  async load() {
    if (!fs.existsSync(this.xuDir)) {
      fs.mkdirSync(this.xuDir);
    } else if (fs.existsSync(`${this.xuDir}xud.conf`)) {
      const configText = fs.readFileSync(`${this.xuDir}xud.conf`);
      try {
        const props = toml.parse(configText);

        // copy parsed json properties from config file to this config object
        util.copyObject(props, this);
      } catch (e) {
        throw new Error(`Parsing error on line ${e.line}, column ${e.column
        }: ${e.message}`);
      }
    }

    if (this.args) {
      // override our config file with command line arguments
      util.copyObject(this.args, this);
      this.args = null;
    }
  }

  get(key) {
    return this.props[key];
  }

  isLndEnabled() {
    return this.swapProtocols[enums.swapProtocols.LND].enable;
  }

  isRaidenEnabled() {
    return this.swapProtocols[enums.swapProtocols.RAIDEN].enable;
  }
}

module.exports = Config;
