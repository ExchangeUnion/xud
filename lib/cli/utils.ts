import colors from 'colors/safe';
import { accessSync, watch } from 'fs';
import os from 'os';
import path from 'path';

const SATOSHIS_PER_COIN = 10 ** 8;

export function getDefaultCertPath() {
  switch (os.platform()) {
    case 'win32': {
      const homeDir = process.env.LOCALAPPDATA!;
      return path.join(homeDir, 'Xud', 'tls.cert');
    }
    case 'darwin': {
      const homeDir = process.env.HOME!;
      return path.join(homeDir, '.xud', 'tls.cert');
    }
    default: {
      const homeDir = process.env.HOME!;
      return path.join(homeDir, '.xud', 'tls.cert');
    }
  }
}

export function shorten(key: string, length = 10) {
  if (key.length <= length * 2 + 3) {
    return key;
  }
  return `${key.slice(0, length)}...${key.slice(key.length - length)}`;
}

export function trim(key: string, length = 10) {
  if (key.length <= length + 3) {
    return '';
  }
  return `${key.slice(0, length)}...`;
}

export const generateHeaders = (headers: string[]) => {
  return headers.map((header) => {
    return colors.blue(header);
  });
};

/** Returns a number of coins as an integer number of satoshis. */
export const coinsToSats = (coinsQuantity: number) => {
  return Math.round(coinsQuantity * SATOSHIS_PER_COIN);
};

/** Returns a number of satoshis as a string representation of coins with up to 8 decimal places. */
export const satsToCoinsStr = (satsQuantity: number) => {
  return (satsQuantity / SATOSHIS_PER_COIN).toFixed(8).replace(/\.?0+$/, '');
};

/** Waits up to 5 seconds for the tls.cert file to be created in case this is the first time xud has been run. */
export const waitForCert = (certPath: string) => {
  return new Promise<void>((resolve, reject) => {
    try {
      accessSync(certPath);
      resolve();
    } catch (err) {
      if (err.code === 'ENOENT') {
        const certDir = path.dirname(certPath);
        const certFilename = path.basename(certPath);
        const fsWatcher = watch(certDir, (event, filename) => {
          if (event === 'change' && filename === certFilename) {
            clearTimeout(timeout);
            fsWatcher.close();
            resolve();
          }
        });
        const timeout = setTimeout(() => {
          fsWatcher.close();
          reject(`timed out waiting for cert to be created at ${certPath}`);
        }, 5000);
      } else {
        // we handle errors due to file not existing, otherwise reject
        reject(err);
      }
    }
  });
};

export function showSeed(seedMnemonicList: string[]) {
  const WORDS_PER_ROW = 4;
  const numberedMnemonic = seedMnemonicList.map((value, index) => {
    return `${index >= 9 ? '' : ' '}${index + 1}. ${value.padEnd(10)}`;
  });
  console.log('----------------------BEGIN XUD SEED---------------------');
  for (let n = 0; n < seedMnemonicList.length / WORDS_PER_ROW; n += 1) {
    console.log(numberedMnemonic.slice(n * WORDS_PER_ROW, (n + 1) * WORDS_PER_ROW).join(' '));
  }
  console.log('-----------------------END XUD SEED----------------------\n');

  console.log(`
Please write down your 24 word mnemonic. It will allow you to recover your xud \
node key and on-chain funds for the initialized wallets listed above should you \
forget your password or lose your device. Off-chain funds in channels can NOT \
be recovered with it and must be backed up and recovered separately. Keep it \
somewhere safe, it is your ONLY backup in case of data loss.
`);
}
