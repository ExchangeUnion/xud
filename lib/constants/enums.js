const enums = exports;

enums.swapProtocols = {
  LND: 'LND',
  RAIDEN: 'RAIDEN',
};

enums.orderingDirections = {
  DESC: 'DESC',
  ASC: 'ASC',
};
