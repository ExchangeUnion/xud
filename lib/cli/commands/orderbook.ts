import Table, { HorizontalTable } from 'cli-table3';
import colors from 'colors/safe';
import { Arguments, Argv } from 'yargs';
import { OrderBookRequest, OrderBookResponse } from '../../proto/xudrpc_pb';
import { callback, loadXudClient } from '../command';
import { satsToCoinsStr } from '../utils';

const COLUMNS = [19, 19, 19, 19];
const HEADER = [
  { content: colors.green('Buy'), colSpan: 2 },
  { content: colors.red('Sell'), colSpan: 2 },
];
const SECONDARY_HEADER = [colors.green('Quantity'), colors.green('Price'), colors.red('Price'), colors.red('Quantity')];

let precision = 0;

const createTable = () => {
  const table = new Table({ colWidths: COLUMNS }) as HorizontalTable;
  table.push(HEADER);
  table.push(SECONDARY_HEADER);
  return table;
};

const displayOrderbook = (pairId: string, orderbook: OrderBookResponse.Buckets.AsObject) => {
  const table = createTable();
  const { buyBucketsList, sellBucketsList } = orderbook;
  const rowCount = Math.max(buyBucketsList.length, sellBucketsList.length);

  for (let n = 0; n < rowCount; n += 1) {
    const row: string[] = [];
    row[0] = buyBucketsList[n] ? satsToCoinsStr(buyBucketsList[n].quantity) : '';
    row[1] = buyBucketsList[n]?.price.toString() ?? '';
    if (row[1] === '0') {
      // instead of displaying zero, display the < symbol and the smallest price bucket
      // given the specified precision. for example if precision is 2, a zero value
      // would be converted to "<0.01"
      row[1] = `<${1 / 10 ** precision}`;
    }
    row[2] = sellBucketsList[n]?.price.toString() ?? '';
    row[3] = sellBucketsList[n] ? satsToCoinsStr(sellBucketsList[n].quantity) : '';
    table.push(row);
  }

  console.log(colors.underline(colors.bold(`\nTrading pair: ${pairId}`)));
  console.log(table.toString());
};

export const displayTables = (orderbooks: OrderBookResponse.AsObject) => {
  orderbooks.bucketsMap.forEach((val) => {
    displayOrderbook(val[0], val[1]);
  });
};

export const command = 'orderbook [pair_id] [precision]';

export const describe = 'display the order book, with orders aggregated per price point';

export const builder = (argv: Argv) =>
  argv
    .option('pair_id', {
      describe: 'trading pair for which to retrieve the order book',
      type: 'string',
    })
    .option('precision', {
      describe: 'the number of digits following the decimal point',
      type: 'number',
      default: 5,
    })
    .option('limit', {
      describe: 'the number of digits following the decimal point',
      type: 'number',
      default: 8,
    })
    .example('$0 orderbook', 'display the order books for all trading pairs')
    .example('$0 orderbook LTC/BTC', 'display the LTC/BTC order book')
    .example('$0 orderbook LTC/BTC 2', 'display the LTC/BTC order book with 2 decimal precision')
    .example('$0 orderbook --precision 2', 'display the order books for all trading pairs with 2 decimal precision')
    .example(
      '$0 orderbook BTC/USDT -2 --limit 10',
      'display the USDT/BTC order book with up to 10 buckets using hundreds place precision',
    );

export const handler = async (argv: Arguments<any>) => {
  const request = new OrderBookRequest();
  const pairId = argv.pair_id ? argv.pair_id.toUpperCase() : undefined;
  request.setPairId(pairId);
  request.setPrecision(argv.precision);
  precision = argv.precision; // store the precision to a global variable in case we need it for display purposes later
  request.setLimit(argv.limit);
  (await loadXudClient(argv)).orderBook(request, callback(argv, displayTables));
};
