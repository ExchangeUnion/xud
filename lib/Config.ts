import fs from 'fs';
import os from 'os';
import path from 'path';
import toml from 'toml';
import utils from './utils/utils';
import { PoolConfig } from './p2p/Pool';
import { LndClientConfig } from './lndclient/LndClient';
import { RaidenClientConfig } from './raidenclient/RaidenClient';
import { DBConfig } from './db/DB';
import { OrderBookConfig } from './orderbook/OrderBook';

class Config {
  public p2p: PoolConfig;
  public xudir: string;
  public db: DBConfig;
  public testDb: DBConfig;
  public rpc: any;
  public lnd: LndClientConfig;
  public raiden: RaidenClientConfig;
  public orderbook: OrderBookConfig;

  constructor(private args?: object) {
    const platform = os.platform();
    let lndDatadir;
    switch (platform) {
      case 'win32': { // windows
        const homeDir = process.env.LOCALAPPDATA;
        this.xudir = `${homeDir}/Xud/`;
        lndDatadir = `${homeDir}/Lnd/`; // default lnd directory location
        break;
      }
      case 'darwin': { // mac
        const homeDir = process.env.HOME;
        this.xudir = `${homeDir}/.xud/`;
        lndDatadir = `${homeDir}/Library/Application Support/Lnd/`;
        break;
      }
      default: { // linux
        const homeDir = process.env.HOME;
        this.xudir = `${homeDir}/.xud/`;
        lndDatadir = `${homeDir}/.lnd/`;
        break;
      }
    }

    // default configuration
    this.p2p = {
      listen: true,
      port: 8885, // X = 88, U = 85 in ASCII
    };
    this.db = {
      host: 'localhost',
      port: 3306,
      username: 'xud',
      database: 'xud',
      dialect: 'mysql',
    };
    this.testDb = {
      ...this.db,
      database: 'xud_test',
    };
    this.rpc = {
      port: 8886,
    };
    this.lnd = {
      disable: false,
      datadir: lndDatadir,
      certpath: path.join(lndDatadir, 'tls.cert'),
      macaroonpath: path.join(lndDatadir, 'admin.macaroon'),
      host: 'localhost',
      port: 10009,
    };
    this.raiden = {
      disable: false,
      host: 'localhost',
      port: 5001,
    };
    this.orderbook = {
      internalmatching: true,
    };
  }

  public async load() {
    if (!fs.existsSync(this.xudir)) {
      fs.mkdirSync(this.xudir);
    } else if (fs.existsSync(`${this.xudir}xud.conf`)) {
      const configText: any = fs.readFileSync(`${this.xudir}xud.conf`);
      try {
        const props = toml.parse(configText);

        // merge parsed json properties from config file to this config object
        utils.deepMerge(this, props);
      } catch (e) {
        throw new Error(`Parsing error on line ${e.line}, column ${e.column
        }: ${e.message}`);
      }
    }

    if (this.args) {
      // override our config file with command line arguments
      utils.deepMerge(this, this.args);
      this.args = undefined;
    }
  }
}

export default Config;
