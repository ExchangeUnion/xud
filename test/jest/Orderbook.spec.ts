import Config from '../../lib/Config';
import { SwapClientType, XuNetwork } from '../../lib/constants/enums';
import DB from '../../lib/db/DB';
import Logger from '../../lib/Logger';
import NodeKey from '../../lib/nodekey/NodeKey';
import Orderbook from '../../lib/orderbook/OrderBook';
import { OwnLimitOrder, OwnMarketOrder } from '../../lib/orderbook/types';
import Network from '../../lib/p2p/Network';
import Peer from '../../lib/p2p/Peer';
import Pool from '../../lib/p2p/Pool';
import SwapClientManager from '../../lib/swaps/SwapClientManager';
import Swaps from '../../lib/swaps/Swaps';
import { UnitConverter } from '../../lib/utils/UnitConverter';
import { getUnitConverter } from '../utils';

jest.mock('../../lib/db/DB', () => {
  return jest.fn().mockImplementation(() => {
    return {
      models: {
        Order: { create: jest.fn() },
        Trade: { create: jest.fn() },
        Pair: {
          findAll: () => {
            return [
              {
                id: pairId,
                baseCurrency: 'LTC',
                quoteCurrency: 'BTC',
              },
              {
                id: 'BTC/USDT',
                baseCurrency: 'BTC',
                quoteCurrency: 'USDT',
              },
            ];
          },
        },
        Currency: {
          findAll: () => {
            const ltc = { id: 'LTC', swapClient: SwapClientType.Lnd };
            const btc = { id: 'BTC', swapClient: SwapClientType.Lnd };
            const usdt = { id: 'USDT', swapClient: SwapClientType.Connext };
            return [
              { ...ltc, toJSON: () => ltc },
              { ...btc, toJSON: () => btc },
              { ...usdt, toJSON: () => usdt },
            ];
          },
        },
      },
    };
  });
});
const advertisedPairs = ['LTC/BTC', 'ETH/BTC'];
const mockActivatePair = jest.fn();
const mockActivateCurrency = jest.fn();
const mockGetIdentifier = jest.fn(() => 'pubkeyoraddress');
const tokenIdentifiers: any = {
  BTC: 'bitcoin-regtest',
  LTC: 'litecoin-regtest',
};
const mockPeerGetTokenIdentifer = jest.fn((currency: string) => tokenIdentifiers[currency]);
jest.mock('../../lib/p2p/Peer', () => {
  return jest.fn().mockImplementation(() => {
    let currencyActive = false;
    return {
      advertisedPairs,
      activatePair: mockActivatePair,
      activateCurrency: (currency: string) => {
        mockActivateCurrency(currency);
        currencyActive = true;
      },
      disabledCurrencies: new Map(),
      isPairActive: () => false,
      isCurrencyActive: () => currencyActive,
      getTokenIdentifier: mockPeerGetTokenIdentifer,
      getIdentifier: mockGetIdentifier,
    };
  });
});
jest.mock('../../lib/p2p/Pool', () => {
  return jest.fn().mockImplementation(() => {
    return {
      updatePairs: jest.fn(),
      on: jest.fn(),
      removeListener: jest.fn(),
      getNetwork: () => XuNetwork.MainNet,
      getTokenIdentifier: (currency: string) => tokenIdentifiers[currency] as string,
      broadcastOrderInvalidation: jest.fn(),
    };
  });
});
jest.mock('../../lib/swaps/Swaps');
jest.mock('../../lib/swaps/SwapClientManager', () => {
  return jest.fn().mockImplementation(() => {
    return {
      canRouteToPeer: jest.fn().mockReturnValue(true),
      isConnected: jest.fn().mockReturnValue(true),
      checkSwapCapacities: jest.fn(),
      get: jest.fn().mockReturnValue({
        maximumOutboundCapacity: () => Number.MAX_SAFE_INTEGER,
      }),
    };
  });
});
jest.mock('../../lib/Logger');
jest.mock('../../lib/nodekey/NodeKey');
const mockedNodeKey = <jest.Mock<NodeKey>>(<any>NodeKey);

const logger = new Logger({});
logger.trace = jest.fn();
logger.verbose = jest.fn();
logger.debug = jest.fn();
logger.error = jest.fn();
logger.info = jest.fn();
const loggers = {
  global: logger,
  db: logger,
  rpc: logger,
  p2p: logger,
  orderbook: logger,
  lnd: logger,
  connext: logger,
  swaps: logger,
  http: logger,
  service: logger,
};

const localId = '97945230-8144-11e9-beb7-49ba94e5bd74';
const pairId = 'LTC/BTC';

describe('OrderBook', () => {
  let pool: Pool;
  let orderbook: Orderbook;
  let config: Config;
  let db: DB;
  let swaps: Swaps;
  let peer: Peer;
  let swapClientManager: SwapClientManager;
  let network: Network;
  let unitConverter: UnitConverter;

  beforeEach(async () => {
    config = new Config();
    network = new Network(XuNetwork.TestNet);
    peer = new Peer(
      loggers.p2p,
      {
        host: 'localhost',
        port: 9735,
      },
      network,
    );
    peer['nodeState'] = {} as any;
    db = new DB(loggers.db, config.dbpath);
    pool = new Pool({
      config: config.p2p,
      xuNetwork: XuNetwork.SimNet,
      logger: loggers.p2p,
      models: db.models,
      version: '1.0.0',
      nodeKey: new mockedNodeKey(),
    });
    pool.broadcastOrder = jest.fn();
    unitConverter = getUnitConverter();
    swapClientManager = new SwapClientManager(config, loggers, unitConverter, db.models);
    swaps = new Swaps({
      pool,
      unitConverter,
      swapClientManager,
      logger: loggers.swaps,
      models: db.models,
    });
    swapClientManager.addInboundReservedAmount = jest.fn();
    swapClientManager.subtractInboundReservedAmount = jest.fn();
    swapClientManager.addOutboundReservedAmount = jest.fn();
    swapClientManager.subtractOutboundReservedAmount = jest.fn();
    swaps.swapClientManager = swapClientManager;
    orderbook = new Orderbook({
      pool,
      swaps,
      unitConverter,
      thresholds: config.orderthresholds,
      logger: loggers.orderbook,
      models: db.models,
      nomatching: config.nomatching,
      nosanityswaps: config.nosanityswaps,
      nobalancechecks: true,
    });
    await orderbook.init();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('nosanityswaps enabled adds pairs and requests orders', async () => {
    orderbook['nosanityswaps'] = true;
    await orderbook['verifyPeerPairs'](peer);
    expect(mockActivateCurrency).toHaveBeenCalledTimes(2);
    expect(mockActivateCurrency).toHaveBeenCalledWith('BTC');
    expect(mockActivateCurrency).toHaveBeenCalledWith('LTC');
    expect(mockActivatePair).toHaveBeenCalledTimes(1);
    expect(mockActivatePair).toHaveBeenCalledWith(advertisedPairs[0]);
  });

  test('isPeerCurrencySupported returns true for a known currency with matching identifiers', async () => {
    expect(orderbook['isPeerCurrencySupported'](peer, 'BTC')).toStrictEqual(true);
    expect(orderbook['isPeerCurrencySupported'](peer, 'LTC')).toStrictEqual(true);
  });

  test('isPeerCurrencySupported returns false for an unknown currency', async () => {
    expect(orderbook['isPeerCurrencySupported'](peer, 'BCH')).toStrictEqual(false);
  });

  test('isPeerCurrencySupported returns false for a known currency with a mismatching identifier', async () => {
    mockPeerGetTokenIdentifer.mockReturnValue('fakecoin-fakenet');
    expect(orderbook['isPeerCurrencySupported'](peer, 'BTC')).toStrictEqual(false);
    expect(orderbook['isPeerCurrencySupported'](peer, 'LTC')).toStrictEqual(false);
  });

  test('isPeerCurrencySupported returns false for a known currency without a swap client identifier for the peer', async () => {
    mockGetIdentifier.mockReturnValue('');
    expect(orderbook['isPeerCurrencySupported'](peer, 'BTC')).toStrictEqual(false);
    expect(orderbook['isPeerCurrencySupported'](peer, 'LTC')).toStrictEqual(false);
  });

  describe('placeOrder', () => {
    test('market order checks swap clients for insufficient inbound balance using best quoted price', async () => {
      const quantity = 20000000;
      const price = 4000;
      const usdtPairId = 'BTC/USDT';
      const isBuy = false;
      // add order to match with market order
      await orderbook.placeLimitOrder({
        order: {
          quantity,
          price,
          pairId: usdtPairId,
          localId: 'matchingorder',
          isBuy: true,
        },
      });

      orderbook['nobalancechecks'] = false;
      const order: OwnMarketOrder = {
        quantity,
        localId,
        isBuy,
        pairId: usdtPairId,
      };
      swaps.swapClientManager.checkSwapCapacities = jest.fn();

      await orderbook.placeMarketOrder({ order });

      expect(swaps.swapClientManager.checkSwapCapacities).toHaveBeenCalledWith(
        expect.objectContaining({ ...order, price }),
      );
    });

    test('placeLimitOrder adds to order book', async () => {
      const quantity = 10000;
      const price = 0.01;
      const order: OwnLimitOrder = {
        quantity,
        pairId,
        localId,
        price,
        isBuy: false,
      };

      await orderbook.placeLimitOrder({ order });
      expect(orderbook.getOwnOrderByLocalId(localId)).toHaveProperty('localId', localId);
      expect(swaps.swapClientManager.addInboundReservedAmount).toHaveBeenCalledWith('BTC', quantity * price);
      expect(swaps.swapClientManager.addOutboundReservedAmount).toHaveBeenCalledWith('LTC', quantity);
    });

    test('placeLimitOrder immediateOrCancel does not add to order book', async () => {
      const quantity = 10000;
      const order: OwnLimitOrder = {
        quantity,
        pairId,
        localId,
        price: 0.01,
        isBuy: false,
      };
      await orderbook.placeLimitOrder({
        order,
        immediateOrCancel: true,
      });
      expect(() => orderbook.getOwnOrderByLocalId(localId)).toThrow(`order with local id ${localId} does not exist`);
    });

    test('placeLimitOrder with replaceOrderId replaces an order ', async () => {
      pool.broadcastOrderInvalidation = jest.fn();
      pool.broadcastOrder = jest.fn();
      const oldQuantity = 10000;
      const oldPrice = 0.01;
      const newQuantity = 20000;
      const newPrice = 0.02;
      const order: OwnLimitOrder = {
        pairId,
        localId,
        quantity: oldQuantity,
        price: oldPrice,
        isBuy: false,
      };

      const oldOrder = await orderbook.placeLimitOrder({ order });
      expect(orderbook.getOwnOrders(pairId).sellArray.length).toEqual(1);
      expect(orderbook.getOwnOrders(pairId).sellArray[0].quantity).toEqual(oldQuantity);
      expect(orderbook.getOwnOrders(pairId).sellArray[0].price).toEqual(oldPrice);
      expect(orderbook.getOwnOrders(pairId).sellArray[0].localId).toEqual(localId);

      const newOrder = await orderbook.placeLimitOrder({
        order: {
          ...order,
          price: newPrice,
          quantity: newQuantity,
        },
        replaceOrderId: localId,
      });

      expect(orderbook.getOwnOrders(pairId).sellArray.length).toEqual(1);
      expect(orderbook.getOwnOrders(pairId).sellArray[0].quantity).toEqual(newQuantity);
      expect(orderbook.getOwnOrders(pairId).sellArray[0].price).toEqual(newPrice);
      expect(orderbook.getOwnOrders(pairId).sellArray[0].localId).toEqual(localId);

      expect(pool.broadcastOrderInvalidation).toHaveBeenCalledTimes(0);
      expect(pool.broadcastOrder).toHaveBeenCalledTimes(2);
      expect(pool.broadcastOrder).toHaveBeenCalledWith({
        id: newOrder.remainingOrder!.id,
        isBuy: false,
        pairId: 'LTC/BTC',
        price: 0.02,
        quantity: 20000,
        replaceOrderId: oldOrder.remainingOrder!.id,
      });
    });
  });

  test('removeOwnOrder removes entire order if dust would have remained', async (done) => {
    const quantity = 10000;
    const order: OwnLimitOrder = {
      quantity,
      pairId,
      localId,
      price: 0.01,
      isBuy: false,
    };
    const { remainingOrder } = await orderbook.placeLimitOrder({ order });
    expect(remainingOrder!.quantity).toEqual(quantity);

    orderbook.on('ownOrder.removed', (orderPortion) => {
      expect(orderPortion.quantity).toEqual(quantity);
      expect(orderPortion.id).toEqual(remainingOrder!.id);
      expect(orderPortion.pairId).toEqual(pairId);
      done();
    });

    const removedOrder = orderbook['removeOwnOrder']({
      pairId,
      orderId: remainingOrder!.id,
      quantityToRemove: quantity - 1,
    });
    expect(removedOrder.quantity).toEqual(quantity);
  });
});
