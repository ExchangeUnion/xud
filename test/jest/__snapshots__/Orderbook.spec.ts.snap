// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderBook placeOrder insufficient outbound balance throws when balancechecks enabled 1`] = `
Object {
  "code": "3.12",
  "message": "LTC outbound balance of 1 is not sufficient for order amount of 10000",
}
`;
